### PR TITLE
*: tiny refactor code to reduce txn conflict on 'table_cache_meta'

### LINKED ISSUE #32386: Transaction conflict on internal SQL, "mysql.table_cache_meta" table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

core bank testing using table caching 

### 2. What did you expect to see? (Required)

Everything goes well

### 3. What did you see instead (Required)

Found some warning log: 

```
can't retry for update statement 'select lock_type, lease, oldReadLease from mysql.table_cache_meta where tid ...'
```

### 4. What is your TiDB version? (Required)

master eabd78cae24b7232139472b4f01ccf80372a627f

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #32386

Problem Summary:

The warning log is caused by the fact that "select ... for update" can't retry when using optimistic transaction.

### What is changed and how it works?

Tiny code refactor. Two changes mainly:

- Remove 'for update', just use the non-'for update' statement

There is no need to worry about write skew abnormity here, we can just use the default RR (SI) 

- Refactor the renew lease code 

In the past, I make a goroutine along with a channel for the renew lease operation, 
i.e. send the renew operation (func) teller the worker to execute it.

```
type renewLeaseCh chan func()
go do.renewLease()

func (do *Domain) renewLease() {
	defer func() {
		do.wg.Done()
		logutil.BgLogger().Info("renew lease goroutine exited.")
	}()
	for {
		select {
		case <-do.exit:
			close(do.renewLeaseCh)
			return
		case op := <-do.renewLeaseCh:
			op()
		}
	}
}
```

And this pattern to avoid duplicated message: https://github.com/pingcap/tidb/pull/31475/files

This is complex than it should be.

To avoid duplicated message, a chan(1) can be used as a token limiter, and spawn a new goroutine is handy:

```
c.renewReadLease = make(chan struct{}, 1)

select {
case c.renewReadLease <- struct{}{}:
	go c.renewLease(ts, data, leaseDuration)
default:
}
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Tiny refactor

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
Ref https://github.com/pingcap/tidb/issues/25293

**Review by tangenta — APPROVED** [review]

File: table/tables/cache.go [hunk 1]
```diff
@@ -109,19 +106,21 @@ func (c *cachedTable) TryReadFromCache(ts uint64, leaseDuration time.Duration) k
...
 // Init is an extra operation for cachedTable after TableFromMeta,
 // Because cachedTable need some additional parameter that can't be passed in TableFromMeta.
-func (c *cachedTable) Init(renewCh chan func(), exec sqlexec.SQLExecutor) error {
-	c.renewCh = renewCh
+func (c *cachedTable) Init(exec sqlexec.SQLExecutor) error {
 	raw, ok := exec.(sqlExec)
 	if !ok {
 		return errors.New("Need sqlExec rather than sqlexec.SQLExecutor")
 	}
+	raw.ExecuteInternal(context.Background(), "set @@session.tidb_retry_limit = 0")
```
Comment (tangenta) [inline table/tables/cache.go:123]: We need to check the error to pass the linter.

**Comment by tangenta** [comment]
Please make sure this is expected: 

Previously each domain can have at most 10 goroutines for renewing lease. After this PR, each cached table in infoSchema can spawn a new goroutine to renew lease.

File: table/tables/cache.go (hunk 1 above)
Comment (tiancaiamao) [reply table/tables/cache.go]: OK, I've removed this line...
Let it retry now, because without retry, there will be some unexpected [WARN] log...

**Comment by tiancaiamao** [comment]
> Please make sure this is expected:
> 
> Previously each domain can have at most 10 goroutines for renewing lease. After this PR, each cached table in infoSchema can spawn a new goroutine to renew lease.

This is by design.
In the old code, there will be duplicated renew lease operation, because there are 10 background goroutines, they can handle several message simultaneously.
Now there will be only one active goroutine.

File: table/tables/cache.go [hunk 2]
```diff
@@ -51,15 +50,14 @@ type cachedTable struct {
...
 	cacheData atomic.Value
 	handle    StateRemote
-	renewCh   chan func()
 	totalSize int64
 
-	mu struct {
-		sync.RWMutex
-		lockingForRead bool
-	}
+	lockingForRead tokenLimit
```
Comment (lcwangchao) [inline table/tables/cache.go:55]: It seems to be a semaphore, why not just use `semaphore` in go lib?

Comment (tiancaiamao) [reply table/tables/cache.go:55]: I'm not familiar with `semaphore`, the channel is more simple for me here.

**Review by lcwangchao — APPROVED** [review]

**Pull request merged**